icu4c/source/common/locdispnames.cpp

```diff
@@ -384,6 +384,13 @@ _getDisplayNameForComponent(const char *locale,
 
     root = tag == _kCountries ? U_ICUDATA_REGION : U_ICUDATA_LANG;
 
+    if (displayLocale != NULL && (*displayLocale == '\0' || *displayLocale == '_')) {
```

Comment:
This is not right.
`*displayLocale == '\0'` is the root locale. For that we want to just load "root".
`*displayLocale == '_'` is any locale starting with an empty language subtag, and it makes no sense to just clobber whatever is in there.